helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -241,11 +244,85 @@ private void refreshClusterConfig(final HelixDataAccessor accessor,
     if (_propertyDataChangedMap.get(HelixConstants.ChangeType.CLUSTER_CONFIG).getAndSet(false)) {
       _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
       refreshedType.add(HelixConstants.ChangeType.CLUSTER_CONFIG);
+      // TODO: This is a temp function to clean up incompatible batched disabled instances format.
+      // Remove in later version.
+      if (needReformatBatchedDIsabledInstance(_clusterConfig) && updateBatchDisableFormat(
+          accessor)) {
+        // read from zkz one more time
+        LogUtil.logInfo(logger, getClusterEventId(), String
+            .format("Clean ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+                getPipelineName()));
+        _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
+      }
       refreshAbnormalStateResolverMap(_clusterConfig);
     } else {
-      LogUtil.logInfo(logger, getClusterEventId(), String.format(
-          "No ClusterConfig change for cluster %s, pipeline %s", _clusterName, getPipelineName()));
+      LogUtil.logDebug(logger, getClusterEventId(), String
+          .format("No ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+              getPipelineName()));
+    }
+  }
+
+  // TODO: This function is used to clean up incompatible batched disabled instances format for
+  // "DISABLED_INSTANCES" introduced in 1.0.3.0. This temp change shoul dbe reverted after 1.0.5.0 \
+  // or later version.
+  private boolean updateBatchDisableFormat(final HelixDataAccessor accessor) {
+    return accessor
+        .updateProperty(accessor.keyBuilder().clusterConfig(), new DataUpdater<ZNRecord>() {
+          @Override
+          public ZNRecord update(ZNRecord currentData) {
+            if (currentData == null) {
+              throw new HelixException(
+                  "Cluster: " + _clusterConfig.getClusterName() + ": cluster config is null");
+            }
+
+            ClusterConfig clusterConfig = new ClusterConfig(currentData);
+            Map<String, String> disabledInstances = clusterConfig.getDisabledInstances();
+            Map<String, String> disabledInstancesWithInfo =
+                clusterConfig.getDisabledInstancesWithInfo();
+
+            ClusterConfig newClusterConfig = new ClusterConfig(currentData);
+            Map<String, String> newDisabledInstances =
+                new TreeMap<>(newClusterConfig.getDisabledInstances());
```

Comment:
Is there any particular reason you used `TreeMap` here? This map and `newDisabledInstancesWithInfo` are only used for put and get, isn't a `HashMap` better for that purpose?

Reply:
TFTR.
This list will be written to ZK. When SRE or other Helix user read the disabled list, having a sorted list is more human readable.

Reply:
I see, this makes sense then. The only thing is, how often is it written to ZK and how large is it usually? I'm just trying to see if it might potentially slow down the pipeline when instances go offline very frequently. If this isn't a concern, then I can resolve the conversation.

Reply:
This is a valid concern. This change will not add a lot load to pipeline because:
1. `refreshClusterConfig` will be a no-op if there is no clusterConfig change callback. clusterConfig change callback will only be triggered in a. the ever first pipeline run or b. there is actual update to clusterConfig.
2. If the format in clusterConfig do not need rewrite (due to one of the following reason: a. The format is clean to begin with, b. the format is already rewritten), `needReformatBatchedDIsabledInstance` will return false and there will be no write back to ZK.
3. This is a temporary clean up fix. This change will be rolled back in next+ version (probably 1.0.5.0).

Reply:
You're right, I had confused this format correction with the general writing of disabled instances. This looks good to me then. Thanks for the explanation.